lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+	}
+	pub(crate) fn is_null(&self) -> bool { self.compact_form[..] == [0; 64][..] }
+	pub(crate) fn empty() -> Self { Self { compact_form: [0; 64] } }
+}
+
+#[repr(C)]
+/// A reference to a serialized transaction, in (pointer, length) form.
+/// This type does *not* own its own memory, so access to it after, eg, the call in which it was
+/// provided to you are invalid.
+pub struct Transaction {
+	pub data: *const u8,
+	pub datalen: usize,
+}
+impl Transaction {
+	pub(crate) fn into_bitcoin(&self) -> BitcoinTransaction {
+		// datalen == 0 is invalid, so don't bother checking here.
+		::bitcoin::consensus::encode::deserialize(unsafe { std::slice::from_raw_parts(self.data, self.datalen) }).unwrap()
+	}
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+}
+
+#[repr(C)]
+#[derive(Clone)]
+/// A transaction output including a scriptPubKey and value.
+/// This type *does* own its own memory, so must be free'd appropriately.
+pub struct TxOut {
+	pub script_pubkey: derived::CVec_u8Z,
+	pub value: u64,
+}
+
+impl TxOut {
+	pub(crate) fn into_rust(self) -> ::bitcoin::blockdata::transaction::TxOut {
+		::bitcoin::blockdata::transaction::TxOut {
+			script_pubkey: self.script_pubkey.into_rust().into(),
+			value: self.value,
+		}
+	}
+	pub(crate) fn from_rust(txout: ::bitcoin::blockdata::transaction::TxOut) -> Self {
+		Self {
+			script_pubkey: CVecTempl::from(txout.script_pubkey.into_bytes()),
+			value: txout.value
+		}
+	}
+}
+#[no_mangle]
+pub extern "C" fn TxOut_free(_res: TxOut) { }
+
+#[repr(C)]
+pub struct u8slice {
+	pub data: *const u8,
+	pub datalen: usize
+}
+impl u8slice {
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[u8] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+pub struct usizeslice {
+	pub data: *const usize,
+	pub datalen: usize
+}
+impl usizeslice {
+	pub(crate) fn from_slice(s: &[usize]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[usize] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// Arbitrary 32 bytes, which could represent one of a few different things. You probably want to
+/// look up the corresponding function in rust-lightning's docs.
+pub struct ThirtyTwoBytes {
+	pub data: [u8; 32],
+}
+impl ThirtyTwoBytes {
+	pub(crate) fn empty() -> Self {
+		Self { data: [0; 32] }
+	}
+}
+
+#[repr(C)]
+pub struct ThreeBytes { pub data: [u8; 3], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct FourBytes { pub data: [u8; 4], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct TenBytes { pub data: [u8; 10], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct SixteenBytes { pub data: [u8; 16], }
```

Comment:
I love how the abundance of proper C arrays let us here LOL

Reply:
Ha, yea, if only C had good support for (fixed-length) arrays.